### PR TITLE
Concurrency implementation using batch processor

### PR DESCRIPTION
**Pull request by berland**
Very much work-in-progress.

### CODE REVIEW

**Comment by berland** [comment]
Code runs through test. Using concurrency on the test workloads leads to slow-down, the penalty of sending dataframes back and forth between Python processes outweighs the benefit from concurrency.

**Comment by berland** [comment]
Self-assessment, concurrent code is missing in `add_from_runpathfile()`, `apply()` and `get_smry()` in the ScratchEnsemble object.

There is no concurrent code in EnsembleSet. Paralellizing this has very little  potential, assuming the there are always more realizations in an ensemble, than ensembles in an ensembleset.

**Comment by asnyv** [comment]
> There is no concurrent code in EnsembleSet. Paralellizing this has very little  potential, assuming the there are always more realizations in an ensemble, than ensembles in an ensembleset.

Flownet might be an exception to that assumption @anders-kiaer ?

**Comment by anders-kiaer** [comment]
> Flownet might be an exception to that assumption @anders-kiaer ?

I actually think @wouterjdb did some test on concurrency earlier, using `flownet` + `fmu-ensemble`?

File: src/fmu/ensemble/__init__.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 except ImportError:
     __version__ = "0.0.0"
 
+from .ensemble import use_concurrent  # noqa
```
Comment (wouterjdb) [inline src/fmu/ensemble/__init__.py:10]: Is this needed?

**Comment by wouterjdb** [comment]
> Self-assessment, concurrent code is missing in `add_from_runpathfile()`, `apply()` and `get_smry()` in the ScratchEnsemble object.
> 
> There is no concurrent code in EnsembleSet. Paralellizing this has very little potential, assuming the there are always more realizations in an ensemble, than ensembles in an ensembleset.

This holds for FlowNet just as well.

File: src/fmu/ensemble/__init__.py (hunk 1 above)
Comment (berland) [reply src/fmu/ensemble/__init__.py:10]: Does not seem so, removed.

**Comment by berland** [comment]
This is now potentially done, except for a big Py27 blocker as it turned out I have used a Py3-only feature (`**somedict`) in a part of the code that is only to be run on Py3. Need a workaround.

File: tests/test_batch.py [hunk 2]
```diff
@@ -86,3 +90,42 @@ def test_yaml():
...
+def test_speedup():
+    """Naive test of speedup in concurrent mode"""
+
+    testdir = os.path.dirname(os.path.abspath(__file__))
+    ens = ScratchEnsemble(
+        "reektest", testdir + "/data/testensemble-reek001/" + "realization-*/iter-0"
+    )
+
+    set_concurrent(True)
+    # really_concurrent = use_concurrent()
```
Comment (asnyv) [inline tests/test_batch.py:113]: Delete? or remove the comment=
```suggestion
```

File: tests/test_batch.py [hunk 3]
```diff
@@ -86,3 +90,42 @@ def test_yaml():
...
+    start_time = datetime.datetime.now()
+    ens.process_batch(batch=[{"apply": {"callback": sleeper}}])
+    end_time = datetime.datetime.now()
+    seq_elapsed = (end_time - start_time).total_seconds()
+    print("FMU_CONCURRENCY: {}".format(use_concurrent()))
+    print("Elapsed time for sequential batch apply sleep: {}".format(seq_elapsed))
+
+    # Can't enforce this, it depends on physical hardware availability.
+    # if really_concurrent:
+    #    assert seq_elapsed > conc_elapsed * 4
```
Comment (asnyv) [inline tests/test_batch.py:131]: Related to the previous comment
```suggestion
```

File: tests/test_batch.py [hunk 4]
```diff
@@ -86,3 +90,42 @@ def test_yaml():
...
+    """Naive test of speedup in concurrent mode"""
+
+    testdir = os.path.dirname(os.path.abspath(__file__))
+    ens = ScratchEnsemble(
+        "reektest", testdir + "/data/testensemble-reek001/" + "realization-*/iter-0"
+    )
+
+    set_concurrent(True)
+    # really_concurrent = use_concurrent()
+    start_time = datetime.datetime.now()
```
Comment (asnyv) [inline tests/test_batch.py:114]: enough to use `time.time()` and save the `datetime` import? (generally, not just this line)
```suggestion
    start_time = time.time()
```

File: tests/test_batch.py [hunk 5]
```diff
@@ -86,3 +90,42 @@ def test_yaml():
...
+    ens.process_batch(batch=[{"apply": {"callback": sleeper}}])
+    end_time = datetime.datetime.now()
+    conc_elapsed = (end_time - start_time).total_seconds()
```
Comment (asnyv) [inline tests/test_batch.py:117]: with time.time() you also wouldn't have to use total_seconds()
```suggestion
    conc_elapsed = (end_time - start_time).total_seconds()
```

File: tests/test_webviz_subsurface_testdata.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
+"""Testing loading the webviz subsurface testdata."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import datetime
```
Comment (asnyv) [inline tests/test_webviz_subsurface_testdata.py:8]: again, consider the simpler `time` package? (not sure if there is any efficiency gain though, but at least less verbose code)

File: src/fmu/ensemble/ensemble.py [hunk 7]
```diff
@@ -65,7 +73,7 @@ class ScratchEnsemble(object):
             the basename of the Eclipse simulation, relative to RUNPATH.
             Fourth column is not used.
         runpathfilter (str): If supplied, the only the runpaths in
```
[the remark this answers is not in the document]
Comment (berland) [reply src/fmu/ensemble/ensemble.py:75]: fixed

File: src/fmu/ensemble/ensemble.py [hunk 8]
```diff
@@ -263,21 +290,46 @@ def add_from_runpathfile(self, runpath, runpathfilter=None, batch=None):
...
+                        row["runpath"],
+                        index=int(row["index"]),
+                        autodiscovery=False,
+                        find_files=[
+                            row["eclbase"] + ".DATA",
+                            row["eclbase"] + ".UNSMRY",
+                        ],
+                        batch=batch,
+                    ).result()
+                    for _, row in runpath_df.iterrows()
```
[the remark this answers is not in the document]
Comment (berland) [reply src/fmu/ensemble/ensemble.py:314]: Fixed. Unaware of the iterrows vs itertuples speed issue :)

File: src/fmu/ensemble/ensemble.py [hunk 9]
```diff
@@ -938,16 +1018,48 @@ def apply(self, callback, **kwargs):
...
-            # Todo: Avoid copy by concatenatint a dict of dataframes
-            # where realization index is the dict keys.
-            result["REAL"] = realidx
-            results.append(result)
-        return pd.concat(results, sort=False, ignore_index=True)
+
+        # It is tempting to just call process_batch() here, but then we
+        # don't know how to collect the results from this particular
+        # apply() operation (if we enforced nonempty localpath, we could)
+        # > kwargs["calllback"] = callback
```
[the remark this answers is not in the document]
Comment (berland) [reply src/fmu/ensemble/ensemble.py:1026]: Fixed

File: src/fmu/ensemble/ensemble.py [hunk 10]
```diff
@@ -938,16 +1018,48 @@ def apply(self, callback, **kwargs):
...
+                    dframe.assign(REAL=r_idx)
+                    for (r_idx, dframe) in dframes_dict_from_apply.items()
+                ]
+
+        else:
+            dframes_from_apply = []
+            for realidx, realization in self.realizations.items():
+                dframes_from_apply.append(
+                    realization.apply(callback, **kwargs).assign(REAL=realidx)
+                )
```
[the remark this answers is not in the document]
Comment (berland) [reply src/fmu/ensemble/ensemble.py:1061]: Fixed.

File: src/fmu/ensemble/ensemble.py [hunk 11]
```diff
@@ -1454,19 +1566,41 @@ def get_smry(
...
-                cache_eclsum=cache_eclsum,
-                include_restart=include_restart,
-            )
-            dframe.insert(0, "REAL", index)
-            dframe.index.name = "DATE"
-            dflist.append(dframe)
-        if dflist:
-            return pd.concat(dflist, sort=False).reset_index()
+        dframes = []
+        if use_concurrent():
```
[the remark this answers is not in the document]
Comment (berland) [reply src/fmu/ensemble/ensemble.py:1570]: Included some inline comments on why. Short answer: process_batch always returns realizations, while get_smry requires dataframes.

File: tests/test_batch.py (hunk 3 above)
Comment (berland) [reply tests/test_batch.py:131]: removed..

File: tests/test_batch.py (hunk 4 above)
Comment (berland) [reply tests/test_batch.py:114]: Fair enough, fixed everywhere.

File: tests/test_batch.py (hunk 5 above)
Comment (berland) [reply tests/test_batch.py:117]: Yes

File: tests/test_webviz_subsurface_testdata.py (hunk 6 above)
Comment (berland) [reply tests/test_webviz_subsurface_testdata.py:8]: removd.

File: tests/test_batch.py (hunk 2 above)
Comment (berland) [reply tests/test_batch.py:113]: removed.

**Comment by berland** [comment]
`get_volumetric_rates()` needs parallelization.

File: src/fmu/ensemble/ensemble.py [hunk 12]
```diff
@@ -199,13 +210,29 @@ def add_realizations(
...
+                loaded_reals = [
+                    executor.submit(
+                        ScratchRealization,
+                        realdir,
+                        realidxregexp=realidxregexp,
+                        autodiscovery=autodiscovery,
+                        batch=batch,
+                    ).result()
+                    for realdir in globbedpaths
+                ]
```
Comment (asnyv) [inline src/fmu/ensemble/ensemble.py:224]: Not really concurrent, as the `.result()` waits for the subprocess to finish for each iteration. Should rather do something like below (which is how it is done in most of the rest of this PR, so this is probably just something you forgot to fix).
```suggestion
               reals_futures = [
                    executor.submit(
                        ScratchRealization,
                        realdir,
                        realidxregexp=realidxregexp,
                        autodiscovery=autodiscovery,
                        batch=batch,
                    )
                    for realdir in globbedpaths
                ]
                loaded_reals = [x.result() for x in reals_futures]
```

File: src/fmu/ensemble/ensemble.py [hunk 13]
```diff
@@ -263,21 +290,43 @@ def add_from_runpathfile(self, runpath, runpathfilter=None, batch=None):
...
+                    executor.submit(
+                        ScratchRealization,
+                        row.runpath,
+                        index=int(row.index),
+                        autodiscovery=False,
+                        find_files=[row.eclbase + ".DATA", row.eclbase + ".UNSMRY",],
+                        batch=batch,
+                    ).result()
+                    for row in runpath_df.itertuples()
+                ]
```
Comment (asnyv) [inline src/fmu/ensemble/ensemble.py:312]: Same as above
```suggestion
            with ProcessPoolExecutor() as executor:
                reals_futures = [
                    executor.submit(
                        ScratchRealization,
                        row.runpath,
                        index=int(row.index),
                        autodiscovery=False,
                        find_files=[row.eclbase + ".DATA", row.eclbase + ".UNSMRY",],
                        batch=batch,
                    )
                    for row in runpath_df.itertuples()
                ]
                ```suggestion
               reals_futures = [
                    executor.submit(
                        ScratchRealization,
                        realdir,
                        realidxregexp=realidxregexp,
                        autodiscovery=autodiscovery,
                        batch=batch,
                    )
                    for realdir in globbedpaths
                ]
                loaded_reals = [x.result() for x in reals_futures]
```
```

File: src/fmu/ensemble/realization.py [hunk 14]
```diff
@@ -945,6 +969,13 @@ def get_eclsum(self, cache=True, include_restart=True):
             EclSum: object representing the summary file. None if
                 nothing was found.
         """
+        # pylint: disable=import-outside-toplevel
+        from .common import use_concurrent
+
+        if use_concurrent():
+            # In concurrent mode, caching is not used as
+            # we do not pickle the loaded EclSum objects
+            cache = False
```
Comment (asnyv) [inline src/fmu/ensemble/realization.py:978]: This change is significant. Each call to `ScratchRealization.get_eclsum()` is quite heavy, and many methods are written with several calls to `get_eclsum()`, such that the loss of caching might actually increase runtime even with concurrency.

File: tests/test_webviz_subsurface_testdata.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
+"""Testing loading the webviz subsurface testdata."""
+
+from __future__ import absolute_import
```
Comment (oysteoh) [inline tests/test_webviz_subsurface_testdata.py:3]: Do we need this now when we are python3 only?

Comment (berland) [reply tests/test_webviz_subsurface_testdata.py:3]: No, it can go away.

**Comment by ertomatic** [comment]
Can one of the admins verify this patch?

**Comment by berland** [comment]
Rebased on top of #182

File: tests/test_webviz_subsurface_testdata.py (hunk 15 above)
Comment (berland) [reply tests/test_webviz_subsurface_testdata.py]: removed.

File: src/fmu/ensemble/common.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        bool: True if concurrency mode should be used
+    """
+    if ENV_NAME not in os.environ:
+        return True
+    env_var = os.environ[ENV_NAME]
+    if (
+        str(env_var) == "0"
+        or str(env_var).lower() == "false"
+        or str(env_var).lower() == "no"
+    ):
```
Comment (oysteoh) [inline src/fmu/ensemble/common.py:27]: maybe this looks better and reduces the usage of `str(..)` and `lower()` ? 
```
env_var = str(os.environ[ENV_NAME]).lower()
if( env_var == "0" or env_var == "false" or env_var == "no"):
    ..
    ..
```

**Comment by berland** [comment]
Superseded by #206

**Pull request closed without merging**